packages/components/src/duotone-picker/duotone-picker.tsx

```diff
@@ -125,6 +159,7 @@ function DuotonePicker( {
 									newColors.length >= 2
 										? newColors
 										: undefined;
+								// @ts-expect-error TODO: Investigate if this is actually a problem
```

Comment:
There's definitely a type mismatch at some point here. `ColorListPicker` declares that both its `value` and its `onChange` deal with `Array< string | undefined >`, while `DuotonePicker` doesn't accept `undefined` in its types.

My gut instinct would be to:
- understand what actually happened in `DuotonePicker` and `ColorListPicker` when the value was `undefined`
- maintain the same behaviour, while adding more explicit checks (ie. small runtime changes) if necessary (e.g adding a typeguard in `ColorListPicker` against `undefined` values?)

Reply:
Alright, @chad1008 and I took a deeper look at this, and came to the conclusion that in order to fix the TS mismatch _properly_ we'd need to rewrite the component in such a way that enforces `value` and `onChange` to deal with a 2 item array (which kind of relates to [this other convo](https://github.com/WordPress/gutenberg/pull/49060#discussion_r1135444042) ).

A quicker way to address the TS error for this PR, while introducing the least amount of runtime changes possible, could be:

1. (Optional) Slightly tweak `ColorListPicker` types

<details>
<summary>Like this</summary>

```diff
diff --git a/packages/components/src/duotone-picker/color-list-picker/index.tsx b/packages/components/src/duotone-picker/color-list-picker/index.tsx
index 8a65382482..cff79027d0 100644
--- a/packages/components/src/duotone-picker/color-list-picker/index.tsx
+++ b/packages/components/src/duotone-picker/color-list-picker/index.tsx
@@ -75,9 +75,11 @@ function ColorListPicker( {
 					disableCustomColors={ disableCustomColors }
 					enableAlpha={ enableAlpha }
 					onChange={ ( newColor ) => {
-						const newColors = value.slice();
-						newColors[ index ] = newColor;
-						onChange( newColors );
+						onChange( [
+							...value.slice( 0, index ),
+							newColor,
+							...value.slice( index + 1 ),
+						] );
 					} }
 				/>
 			) ) }
diff --git a/packages/components/src/duotone-picker/color-list-picker/types.ts b/packages/components/src/duotone-picker/color-list-picker/types.ts
index becfc405b2..73c1900235 100644
--- a/packages/components/src/duotone-picker/color-list-picker/types.ts
+++ b/packages/components/src/duotone-picker/color-list-picker/types.ts
@@ -21,7 +21,7 @@ export type ColorListPickerProps = {
 	/**
 	 * An array containing the currently selected colors.
 	 */
-	value?: Array< string | undefined >;
+	value?: Array< string >;
 	/**
 	 * Controls whether the custom color picker is displayed.
 	 */

```
</details>

2. Cast the `newColors` array as `string[]`. This is not a 100% correct cast, since in theory `newColors` could still have `undefined` items at index 2 and up, but we kind of know that in practice that's not supposed to happen.

<details>
<summary>Like this</summary>

```diff
diff --git a/packages/components/src/duotone-picker/duotone-picker.tsx b/packages/components/src/duotone-picker/duotone-picker.tsx
index 7d854242e5..4f909b01fe 100644
--- a/packages/components/src/duotone-picker/duotone-picker.tsx
+++ b/packages/components/src/duotone-picker/duotone-picker.tsx
@@ -157,9 +157,8 @@ function DuotonePicker( {
 								}
 								const newValue =
 									newColors.length >= 2
-										? newColors
+										? ( newColors as string[] )
 										: undefined;
-								// @ts-expect-error TODO: Investigate if this is actually a problem
 								onChange( newValue );
 							} }
 						/>

```
</details>

Reply:
Thanks for looking into it, Marco + Chad!

I'm not quite convinced that the `string[]` cast is any better than a `ts-expect-error` — if anything it can imply a false sense of safety. Do you prefer this cast rather than leaving the error for proper fixing later? If not, I can update the TODO comment so it links to this PR thread 😄

Reply:
You make a good point — we can leave the `@ts-expect-error` instead of the typecast to avoid the false sense of safety (plus, we will get a TS error if we ever fix those types!).

What do you think about the tweaks suggested in the first point above?

Reply:
Works for me!

Last thing I'd suggest is to improve the comment associated to the `@ts-expect-error`, explaining that it caused by the fact that typescript doesn't know that both `newColors` and `value` are in fact supposed to be a 2 item tuple, instead of a `string[]`

Reply:
✅ 2c31a5c1